### PR TITLE
[Type checker] Requestify type checking of a function body

### PR DESCRIPTION
**Pull request by DougGregor**
More fully requestify the type checking of a function body, so it covers all uses of "type check a function body (up to a certain point)" in the request. While here, make a few more synthesized bodies lazy:

* Bodies of implicitly-defined class deinitializers
* Bodies of default initializers

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: include/swift/Basic/SourceLoc.h [hunk 1]
```diff
@@ -78,6 +78,14 @@ class SourceLoc {
   }
 
   void dump(const SourceManager &SM) const;
+
+	friend size_t hash_value(SourceLoc loc) {
+		return reinterpret_cast<uintptr_t>(loc.getOpaquePointerValue());
```
Comment (slavapestov) [inline include/swift/Basic/SourceLoc.h]: Weird indentation here

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: include/swift/Basic/SourceLoc.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/Basic/SourceLoc.h]: Huh, tabs. I'll clean it up, thanks!

**Pull request merged**